src/main/java/appeng/me/storage/ItemCellInventory.java

```diff
@@ -0,0 +1,316 @@
+
+
+import java.util.HashSet;
+
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraftforge.oredict.OreDictionary;
+
+import appeng.api.AEApi;
+import appeng.api.config.Actionable;
+import appeng.api.exceptions.AppEngException;
+import appeng.api.implementations.items.IStorageCell;
+import appeng.api.networking.security.IActionSource;
+import appeng.api.storage.IMEInventory;
+import appeng.api.storage.IMEInventoryHandler;
+import appeng.api.storage.ISaveProvider;
+import appeng.api.storage.IStorageChannel;
+import appeng.api.storage.channels.IItemStorageChannel;
+import appeng.api.storage.data.IAEItemStack;
+import appeng.core.AEConfig;
+import appeng.core.AELog;
+import appeng.util.Platform;
+import appeng.util.item.AEItemStack;
+
+
+public class ItemCellInventory extends CellInventoryBase<IAEItemStack>
+{
+	private static final HashSet<Integer> BLACK_LIST = new HashSet<>();
+
+	protected ItemCellInventory( final NBTTagCompound data, final ISaveProvider container )
+	{
+		super( data, container, 8 );
+	}
+
+	private ItemCellInventory( final ItemStack o, final ISaveProvider container ) throws AppEngException
+	{
+		super( o, container, 8 );
+	}
+
+	public static IMEInventoryHandler getCell( final ItemStack o, final ISaveProvider container2 )
+	{
+		try
+		{
+			return new CellInventoryHandler( new ItemCellInventory( o, container2 ), AEApi.instance().storage().getStorageChannel( IItemStorageChannel.class ) );
+		}
+		catch( final AppEngException e )
+		{
+			return null;
+		}
+	}
+
+	private static boolean isStorageCell( final ItemStack i )
+	{
+		if( i == null )
+		{
+			return false;
+		}
+
+		try
+		{
+			final Item type = i.getItem();
+			if( type instanceof IStorageCell )
+			{
+				return !( (IStorageCell) type ).storableInStorageCell();
+			}
+		}
+		catch( final Throwable err )
+		{
+			return true;
+		}
+
+		return false;
+	}
+
+	public static boolean isCell( final ItemStack i )
+	{
+		if( i == null )
+		{
+			return false;
+		}
+
+		final Item type = i.getItem();
+		if( type instanceof IStorageCell )
+		{
+			if ( ( (IStorageCell) type ).getChannel() == AEApi.instance().storage().getStorageChannel( IItemStorageChannel.class ) )
+			{
+				return ( (IStorageCell) type ).isStorageCell( i );
+			}
+		}
+
+		return false;
+	}
+
+	public static void addBasicBlackList( final int itemID, final int meta )
+	{
+		BLACK_LIST.add( ( meta << Platform.DEF_OFFSET ) | itemID );
+	}
+
+	private static boolean isBlackListed( final IAEItemStack input )
+	{
+		if( BLACK_LIST.contains( ( OreDictionary.WILDCARD_VALUE << Platform.DEF_OFFSET ) | Item.getIdFromItem( input.getItem() ) ) )
+		{
+			return true;
+		}
+		return BLACK_LIST.contains( ( input.getItemDamage() << Platform.DEF_OFFSET ) | Item.getIdFromItem( input.getItem() ) );
+	}
+
+	@Override
+	public IAEItemStack injectItems( final IAEItemStack input, final Actionable mode, final IActionSource src )
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') The Cyclomatic Complexity of this method "injectItems" is 18 which is greater than 10 authorized. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)